processing/src/main/java/org/apache/druid/jackson/DefaultObjectMapper.java

```diff
@@ -61,11 +77,54 @@ public DefaultObjectMapper(JsonFactory factory)
     configure(MapperFeature.ALLOW_FINAL_FIELDS_AS_MUTATORS, false);
     configure(SerializationFeature.INDENT_OUTPUT, false);
     configure(SerializationFeature.FLUSH_AFTER_WRITE_VALUE, false);
+
+    addHandler(new DefaultDeserializationProblemHandler(serviceName));
   }
 
   @Override
   public ObjectMapper copy()
   {
     return new DefaultObjectMapper(this);
   }
+
+  /**
+   * A custom implementation of {@link #DeserializationProblemHandler} to add custom error message so
+   * that users know how to troubleshoot unknown type ids.
+   */
+  static class DefaultDeserializationProblemHandler extends DeserializationProblemHandler
+  {
+    private final String serviceName;
+
+    public DefaultDeserializationProblemHandler(@Nullable String serviceName)
+    {
+      this.serviceName = serviceName == null ? "unknown" : serviceName;
+    }
+
+    @VisibleForTesting
+    String getServiceName()
+    {
+      return serviceName;
+    }
+
+    @Override
+    public JavaType handleUnknownTypeId(DeserializationContext ctxt,
+                                        JavaType baseType, String subTypeId, TypeIdResolver idResolver,
+                                        String failureMsg)
+        throws IOException
+    {
+      String msg = String.format("Please make sure to load all the necessary extensions and jars " +
+              "with type '%s' on '%s' service. " +
+              "Could not resolve type id '%s' as a subtype of %s",
```

Comment:
Super nit: I think it's better to state the problem first and then a possible resolution.

```suggestion
      String msg = String.format("Could not resolve type id '%s' as a subtype of %s. " +
              "Please ensure that the extensions and jars defining the subtype '%s' " +
              "are loaded on service '%s'.\n\n\n",
```

Also maybe add trailing newlines so that the appended `failureMsg` does not obscure the error message we want to communicate.